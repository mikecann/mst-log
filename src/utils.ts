export const exhaustiveCheck = (param: never): never => {
  throw new Error("should not reach here");
};

export const repeat = (str: string, times: number) => new Array(times + 1).join(str);

export const pad = (num: number, maxLength: number) =>
  repeat("0", maxLength - num.toString().length) + num;

export const formatTime = (nowMs: number) => {
  const time = new Date(nowMs);
  return `${pad(time.getHours(), 2)}:${pad(time.getMinutes(), 2)}:${pad(
    time.getSeconds(),
    2
  )}.${pad(time.getMilliseconds(), 3)}`;
};

export const getTime = () => {
  const timer =
    typeof performance !== "undefined" &&
    performance !== null &&
    typeof performance.now === "function"
      ? performance
      : Date;

  return timer.now();
};
