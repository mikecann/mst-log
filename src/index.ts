import { IMiddlewareEvent, getPath, getSnapshot, IMiddlewareEventType } from "mobx-state-tree";
import { exhaustiveCheck, formatTime, getTime } from "./utils";

type Parentable = {
  parentEvent: Parentable | undefined;
  name: string;
};

const getCallName = (call: Parentable): string =>
  call.parentEvent ? getCallName(call.parentEvent) : call.name;

export type Logger = {
  group: typeof console.group;
  groupCollapsed: typeof console.groupCollapsed;
  groupEnd: typeof console.groupEnd;
  log: typeof console.log;
};

export type Options = {
  getCallTypeColor: (type: IMiddlewareEventType) => string;
  getShouldGroupBeCollapsed: (call: IMiddlewareEvent) => boolean;
  logger: Logger;
  getTime: () => number;
};

export const mstLog = (options?: Partial<Options>) => {
  // Choosing either the provided options or the default values
  const _getCallTypeColor = (options && options.getCallTypeColor) ?? getCallTypeColor;
  const _logger = (options && options.logger) ?? console;
  const _getShouldGroupBeCollapsed =
    (options && options.getShouldGroupBeCollapsed) ?? getShouldGroupBeCollapsed;
  const _getTime = (options && options.getTime) ?? getTime;

  // the actual middleware
  return (call: IMiddlewareEvent, next: (call: IMiddlewareEvent) => void) => {
    const snapshotBefore = getSnapshot(call.tree);
    const timeBefore = _getTime();

    const grp = _getShouldGroupBeCollapsed(call) ? _logger.groupCollapsed : _logger.group;
    grp(
      [
        `%c ${call.type}`,
        `%c${getPath(call.context)}/${getCallName(call)}`,
        `%c@ ${formatTime(timeBefore)}`,
      ].join(" "),
      ...[
        `color: ${_getCallTypeColor(call.type)}; font-weight: lighter;`,
        `color: inherit; font-weight: bold;`,
        `color: gray; font-weight: lighter;`,
      ]
    );

    _logger.log(`%c args`, `color: grey; font-weight: bold;`, ...call.args);
    _logger.log(`%c prev state`, `color: #cc8de0; font-weight: bold;`, snapshotBefore);

    next(call);

    const timeAfter = _getTime();
    const snapshotAfter = getSnapshot(call.tree);

    _logger.log(`%c next state`, `color: #4CAF50; font-weight: bold;`, snapshotAfter);
    _logger.log(`%c ${(timeAfter - timeBefore).toFixed(2)}ms elapsed`, `color: grey;`);
    _logger.groupEnd();
  };
};

const getCallTypeColor = (type: IMiddlewareEventType): string => {
  if (type == "action") return `#74cccf`;
  if (type == "flow_spawn") return `#74b4cf`;
  if (type == "flow_resume") return `#cfcf74`;
  if (type == "flow_return") return `#74cf8b`;
  if (type == "flow_resume_error") return `#cf7474`;
  if (type == "flow_throw") return `#c91c1c`;
  return exhaustiveCheck(type);
};

const getShouldGroupBeCollapsed = (call: IMiddlewareEvent) =>
  call.type == "flow_resume_error" ? false : true;
