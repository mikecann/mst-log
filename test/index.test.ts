import { types, flow, addMiddleware, IMiddlewareEventType } from "mobx-state-tree";
import { mstLog, Logger } from "../src/index";
import { recordingLogger, TestStore } from "./helpers";
import { exhaustiveCheck } from "../src/utils";

it("can log a tree", async () => {
  const { logger, getHistory } = recordingLogger();
  const store = TestStore.create({});

  addMiddleware(
    store,
    mstLog({
      logger,
      getTime: () => 0,
    })
  );

  store.simpleSetAction({ a: 123, b: { c: "hi" } });
  await store.asyncAction(true);

  try {
    store.errorThrowingAction();
  } catch (e) {}

  try {
    await store.errorThrowingAsync("some input");
  } catch (e) {}

  expect(getHistory()).toMatchSnapshot();
});

it("can have the colors changed via the options", async () => {
  const { logger, getHistory } = recordingLogger();
  const store = TestStore.create({});

  addMiddleware(
    store,
    mstLog({
      logger,
      getTime: () => 0,
      getCallTypeColor: (type: IMiddlewareEventType): string => {
        if (type == "action") return `#000001`;
        if (type == "flow_spawn") return `#000002`;
        if (type == "flow_resume") return `#000003`;
        if (type == "flow_return") return `#000004`;
        if (type == "flow_resume_error") return `#000005`;
        if (type == "flow_throw") return `#000006`;
        return exhaustiveCheck(type);
      },
    })
  );

  store.simpleSetAction({ a: 123, b: { c: "hi" } });
  await store.asyncAction(true);

  try {
    store.errorThrowingAction();
  } catch (e) {}

  try {
    await store.errorThrowingAsync("some input");
  } catch (e) {}

  expect(getHistory()).toMatchSnapshot();
});

it("can the logging group collapsed'ness changed via the options", async () => {
  const { logger, getHistory } = recordingLogger();
  const store = TestStore.create({});

  let shouldBeCollapsed = false;

  addMiddleware(
    store,
    mstLog({
      logger,
      getTime: () => 0,
      getShouldGroupBeCollapsed: () => shouldBeCollapsed,
    })
  );

  store.simpleSetAction({ a: 123, b: { c: "hi" } });

  shouldBeCollapsed = true;

  store.simpleSetAction({ a: 456, b: { c: "hi" } });

  expect(getHistory()).toMatchSnapshot();
});

it("can output the elapsed time", async () => {
  const { logger, getHistory } = recordingLogger();
  const store = TestStore.create({});

  let times = [10000000, 10010000, 10020000, 10030000, 10040000];
  let index = 0;

  addMiddleware(
    store,
    mstLog({
      logger,
      getTime: () => times[index++],
    })
  );

  store.simpleSetAction({ a: 123, b: { c: "hi" } });

  expect(getHistory()).toMatchSnapshot();
});
