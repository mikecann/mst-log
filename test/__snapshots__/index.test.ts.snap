// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can have the colors changed via the options 1`] = `
Array [
  "groupCollapsed - [\\"%c action %c/simpleSetAction %c@ 08:00:00.000\\",\\"color: #000001; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",{\\"a\\":123,\\"b\\":{\\"c\\":\\"hi\\"}}]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c action %c/asyncAction %c@ 08:00:00.000\\",\\"color: #000001; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",true]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c flow_spawn %c/asyncAction %c@ 08:00:00.000\\",\\"color: #000002; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",true]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c flow_resume %c/asyncAction %c@ 08:00:00.000\\",\\"color: #000003; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",null]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c flow_resume %c/asyncAction %c@ 08:00:00.000\\",\\"color: #000003; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",\\"aaa\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c flow_resume %c/asyncAction %c@ 08:00:00.000\\",\\"color: #000003; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",123]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c flow_return %c/asyncAction %c@ 08:00:00.000\\",\\"color: #000004; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",\\"foo\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c action %c/errorThrowingAction %c@ 08:00:00.000\\",\\"color: #000001; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c action %c/errorThrowingAsync %c@ 08:00:00.000\\",\\"color: #000001; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",\\"some input\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c flow_spawn %c/errorThrowingAsync %c@ 08:00:00.000\\",\\"color: #000002; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",\\"some input\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c flow_resume %c/errorThrowingAsync %c@ 08:00:00.000\\",\\"color: #000003; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",null]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c flow_resume %c/errorThrowingAsync %c@ 08:00:00.000\\",\\"color: #000003; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",\\"aaa\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c flow_throw %c/errorThrowingAsync %c@ 08:00:00.000\\",\\"color: #000006; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",{}]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
]
`;

exports[`can log a tree 1`] = `
Array [
  "groupCollapsed - [\\"%c action %c/simpleSetAction %c@ 08:00:00.000\\",\\"color: #74cccf; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",{\\"a\\":123,\\"b\\":{\\"c\\":\\"hi\\"}}]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c action %c/asyncAction %c@ 08:00:00.000\\",\\"color: #74cccf; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",true]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c flow_spawn %c/asyncAction %c@ 08:00:00.000\\",\\"color: #74b4cf; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",true]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c flow_resume %c/asyncAction %c@ 08:00:00.000\\",\\"color: #cfcf74; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",null]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c flow_resume %c/asyncAction %c@ 08:00:00.000\\",\\"color: #cfcf74; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",\\"aaa\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c flow_resume %c/asyncAction %c@ 08:00:00.000\\",\\"color: #cfcf74; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",123]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c flow_return %c/asyncAction %c@ 08:00:00.000\\",\\"color: #74cf8b; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",\\"foo\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c action %c/errorThrowingAction %c@ 08:00:00.000\\",\\"color: #74cccf; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c action %c/errorThrowingAsync %c@ 08:00:00.000\\",\\"color: #74cccf; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",\\"some input\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c flow_spawn %c/errorThrowingAsync %c@ 08:00:00.000\\",\\"color: #74b4cf; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",\\"some input\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c flow_resume %c/errorThrowingAsync %c@ 08:00:00.000\\",\\"color: #cfcf74; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",null]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c flow_resume %c/errorThrowingAsync %c@ 08:00:00.000\\",\\"color: #cfcf74; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",\\"aaa\\"]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "groupCollapsed - [\\"%c flow_throw %c/errorThrowingAsync %c@ 08:00:00.000\\",\\"color: #c91c1c; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",{}]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
]
`;

exports[`can output the elapsed time 1`] = `
Array [
  "groupCollapsed - [\\"%c action %c/simpleSetAction %c@ 10:46:40.000\\",\\"color: #74cccf; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",{\\"a\\":123,\\"b\\":{\\"c\\":\\"hi\\"}}]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 10000.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
]
`;

exports[`can the logging group collapsed'ness changed via the options 1`] = `
Array [
  "group - [\\"%c action %c/simpleSetAction %c@ 08:00:00.000\\",\\"color: #74cccf; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",{\\"a\\":123,\\"b\\":{\\"c\\":\\"hi\\"}}]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
  "groupCollapsed - [\\"%c action %c/simpleSetAction %c@ 08:00:00.000\\",\\"color: #74cccf; font-weight: lighter;\\",\\"color: inherit; font-weight: bold;\\",\\"color: gray; font-weight: lighter;\\"]",
  "log - [\\"%c args\\",\\"color: grey; font-weight: bold;\\",{\\"a\\":456,\\"b\\":{\\"c\\":\\"hi\\"}}]",
  "log - [\\"%c prev state\\",\\"color: #cc8de0; font-weight: bold;\\",{}]",
  "log - [\\"%c next state\\",\\"color: #4CAF50; font-weight: bold;\\",{}]",
  "log - [\\"%c 0.00ms elapsed\\",\\"color: grey;\\"]",
  "groupEnd - []",
]
`;
